### DISCUSSION TITLE
validate_associated plugin errors from many_to_many association

### DISCUSSION THREAD

**Discussion by opya** [jeremyevans/sequel#2123] [General]
Hello,

I have a model Card that has many_to_many association with the Order model. In the Card model, i have a method similar to the one in the plugin examples [page](https://sequel.jeremyevans.net/rdoc-plugins/classes/Sequel/Plugins/ValidateAssociated.html) something like this:

```ruby
Sequel::Model.plugin :validate_associated

class Card < Sequel::Model
  many_to_many :orders
  
  def check_orders!
    orders.each do |order|
      delay_validate_associated_object(model.association_reflection(:orders), order)
    end
  end
end

class Order < Sequel::Model
end
```

suppose i have a Card instance card, when using card.check_orders! and then valid?  i get something like this as an errors

```ruby
{:orders=>["delivery_address Delivery address is missing."]}
```

i was wondering if there a way to get a result similar to this:
```ruby
{:orders=>{ delivery_address: "Delivery address is missing."} }
```

i've checked here https://github.com/jeremyevans/sequel/blob/e296f6e5de8750c59b067a69e495ab1890f7b353/lib/sequel/plugins/validate_associated.rb#L75-L77

and i believe that this will work for me, but will change how the plugin works currently

```diff
+errors.add(association, obj.errors)
-obj.errors.full_messages.each do |m|
-    errors.add(association, m)
-end
```

as always, i'm sure there is a proper way to do that in Sequel :) maybe i missed some plugin or option. I'll be glad if someone helps me out.

**Comment by jeremyevans** [comment]
Model::Errors values are currently expected to be strings, so Sequel cannot currently support `{:orders=>{ delivery_address: "Delivery address is missing."} }`.  Potentially, this could be supported via a plugin that subclasses Model::Errors and added support for nested errors.

**Reply by jeremyevans** [reply]
To clarify, Model::Errors values are expected to be arrays of strings (to support multiple error messages per column).

**Reply by opya** [reply]
checked the Model::Errors, but didn't get the part that you suggested that we can subclass the Model::Errors and extract it as a plugin. Checked how the other plugins are made and all of them are modules. Probably i don't get the whole picture at all, can you give me one more joker, please

**Reply by jeremyevans** [reply]
You can override Sequel::Model#errors_class in a plugin to specify a separate errors class (which could subclass from Sequel::Model::Errors), and then load that plugin into models where you want to support nested errors.
